### DISCUSSION TITLE
Dynamically Rotate Image to Horizontal Alignment

### DISCUSSION THREAD

**Discussion by lm-sudo** [ImageMagick/ImageMagick#6856] [Help] [answered]
### ImageMagick version

6.9.10-23

### Operating system, version and so on

Ubuntu 20.04.6 LTS

### Description

Hello,

My conditions and goal are as follows:

1. A label will be rested upon a colored background(think of a 4x2 barcode with some text against a backdrop).
2. An image of the label will be taken with a camera(photo), there is no way to guarantee any kind of orientation. That is, the image may be taken upside down, diagonally, horizontally, at a skew, etc. The photo cannot be taken with a label edge side-on such that the contents of the label are not visible.
3. The EXIF orientation meta data is unknown because the device to be used is unknown, it is possible that it will exist and be correct, or it may not exist, or it may exist and be incorrect(as I've read is sometimes the case with Samsung devices).
4. The background color is standardized after being taken such that the background is all one uniform hex color with no deviations.
5. The label needs to be rotated (the direction of rotation is to my knowledge, not relevant) such that the label becomes horizontally aligned.
6. It is okay if the image is mirrored or upside down, but it must be horizontal. 

Note: This is one step in a larger goal I have, which is to produce a cropped image of nothing but the label, have it horizontally aligned, and right side up. 

Some pictures for visual aid/example:
![example1](https://github.com/ImageMagick/ImageMagick/assets/89110072/3d7e0255-fa08-4fa6-86be-8fcf1ac3e7b6)
![example2](https://github.com/ImageMagick/ImageMagick/assets/89110072/487e2d6e-15ae-4767-aac0-49cabd4c3d86)
![example3](https://github.com/ImageMagick/ImageMagick/assets/89110072/022c11a3-7315-4716-8ad0-92941ca2d597)
![goal](https://github.com/ImageMagick/ImageMagick/assets/89110072/f5351603-9d0d-4547-8aab-05bc26a58ee8)

My Current plan:
1.  Define center coordinates of the label (x_1 , y_1).
2.  Define dimensions of the label (L by W).
3. Scan page starting at the top, from left to right and determine first non "color" pixel(black or white).
4. Define coordinates of first non "color" pixel (x_2 , y_2).
5. Define reference point on coordinates by making a copy of coordinates defined in step 4 and then setting x-coordinate = 0 (x_3 , y_3).
6. Rotate image counter-clockwise until y_2 - y_1 = W x 1/2.

![my_plan](https://github.com/ImageMagick/ImageMagick/assets/89110072/57161e73-954e-4027-9e21-41a330d3da75)

Help:
Is there a simpler or cleaner or more consistent way to do this using image magick, php, or python? If so, please explain.

Will my current plan work, or do you foresee some problems with this method?

If you think my current plan will work, do you know of any ways to scan an image starting at the top, from left to right, and stopping at the first non "color" pixel?

Thanks in advance for any help you can provide.

**Comment by fmw42** [comment]
Please clarify. Is the background rotated with the label?  Or do you want to extract the label and rotate it and possibly put it back on a similar background?

You can use Imagemagick rotated rectangle processing find the shape and orientation of the label. Then rotate using that. 
See minimum bounding box and unrotated at https://imagemagick.org/script/convex-hull.php.  You should threshold your image so that the label is white and the background is black. Use morphology to remove small spots and close up the rectangle as much as possible. Then get the rotated rectangle angle from the thresholded image. Then apply the rotation to the original image.

Python/OpenCV also has cv2.minAreaRect(), which is similar.

**Reply by lm-sudo** [reply]
I would like extract the label and rotate it, I then intend to set it on top of a pdf. I do not want to use the background, I have added it in the hopes it will make extraction of the label easier for any processing tools. I will look into the documentation for the tools you've mentioned. I've only been using image magick for about a week so I am extremely unfamiliar with the terminology you're using and will need some time to try out your method. Thank you for the direction.

**Reply by fmw42** [reply]
I will make an example later today. Unfortunately, I have to leave now for a while.

**Reply by fmw42** [reply, marked as answer]
Here is an example.  It is in Unix syntax.  In place of the morphology, I am using connected components processing to extract the largest white region from the threshold.  Imagemagick cannot tell if the text is right side up or up side down. So you may need to apply a 180 degree rotation.

Input:

![label](https://github.com/ImageMagick/ImageMagick/assets/13458218/cd1ecff5-1d67-4435-bd38-db1ad418b4a0)


```
Line 1: read the input
Line 2: copy the input and apply color thresholding to make the label white and the background black
Line 3-6: apply connected components processing on the binary image to remove regions smaller than 100,000 pixels in area so that we have only the white label without holes on perfectly black background
Line 7: put that image in the alpha channel of the input so that it has a transparent background
Line 8: get the minimum bounding rectangle unrotation angle
Line 9: rotate the input image (that has transparent background) and trim to the label
Line 10: save the output

```

```
magick label.jpg \
\( +clone -color-threshold "srgb(140,180,220)-srgb(170,210,255)" \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \) \
-alpha off -compose copy_opacity -composite \
-background none -set option:angle "%[minimum-bounding-box:unrotate]" \
-rotate "%[angle]" -trim +repage \
label_rotated.png

```

Results

![label_rotated](https://github.com/ImageMagick/ImageMagick/assets/13458218/b6a02d6d-3135-469f-a0bb-27bab0f09856)


For connected components processing see https://imagemagick.org/script/connected-components.php

**Reply by lm-sudo** [reply]
Thank you, I will test this on my end shortly. I appreciate your effort/time.

**Reply by lm-sudo** [reply]
Question, are the forward slashes intended to be return characters? Are they perhaps escaping something? I'm not familiar with that convention.

**Reply by fmw42** [reply]
Copy my code exactly.  (You may remove the end of line slashes, but then you have to close the code up to one very long line.)

Forward slashes at the end of the line represent line continuation. This is one long command, but split into multiple lines for easy reading. That is bash shell script syntax. Other end of line characters are used in Windows, i.e. ^.   The slashes before the parentheses are escapes.  This is Imagemagick needed for Unix (not needed for Windows)

**Reply by lm-sudo** [reply]
Script's current form:
`convert /sub1/sub2/sub3/sub4/file.jpg \
\( +clone -color-threshold "srgb(140,180,220)-srgb(170,210,255)" \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \) \
-alpha off -compose copy_opacity -composite \
-background none -set option:angle "%[minimum-bounding-box:unrotate]" \
-rotate "%[angle]" -trim +repage \
/sub1/sub2/sub3/sub4/file.png`

Replaced magick with convert because I was receiving: 
Command 'magick' not found, did you mean:

  command 'magic' from deb magic (8.2.157+ds.1-1)

Replaced label.jpg and label_rotated.png with desired path.

Currently I face this error:
convert-im6.q16: unrecognized option `-color-threshold' @ error/convert.c/ConvertImageCommand/1198.

Tinkering with the script for now.

EDIT: It appears -color-threshold is not available on my version of image magick. Going to see if I can get the version upgraded.

**Reply by fmw42** [reply]
Color-threshold is only available in IM 7. You would have to replace it with -fuzz ZZ% -fill white -opaque XXX -fill black +opaque white so that you get white for the rectangle and black elsewhere. However, the minimum bounding box also is only available for IM 7.  So you will have to upgrade to IM 7 or look to a Python/OpenCV script that does the equivalent.

**Reply by lm-sudo** [reply]
Alright, I managed to spin up a vm running image magick 7.1.1-21. Now I am able to run your code exactly and it works as intended. I fed it this image and it returned this image.
![label3](https://github.com/ImageMagick/ImageMagick/assets/89110072/c9244916-83f5-4115-be63-7af282593cc8)
![label_rotated3](https://github.com/ImageMagick/ImageMagick/assets/89110072/5e0e4369-c40b-4e9a-b1f8-a52171c044a3)
Any idea what may be causing this sort of blending on the barcode? I tested on a few other labels and this happens fairly often.

**Reply by fmw42** [reply]
It likely needs a change to the fuzz value in getting the color threshold and/or and increase the area-threshold in the connected components processing.  Look at the mask.png image. Does the label area have any black inside it? It should not if the fuzz value and area-threshold is adequate.

You can check the color threshold result by adding +write tmp.png after that step

```
convert label.jpg -auto-orient +repage +write img.png \
-fuzz 4.5% -fill white -opaque "srgb(153,192,233)" -fill black +opaque white +write tmp.png \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \
mask.png
cropbox=`convert mask.png -format "%@" info:`
convert mask.png -crop $cropbox +repage mask2.png
convert img.png -crop $cropbox +repage img2.png
angle=`unrotate2 -l 10 mask2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`
echo $angle
declare `convert img2.png -format "aspect_flag=%[fx:w/h<1?1:0]" info:`
[ $aspect_flag -eq 1 ] && angle2=`convert xc: -format "%[fx:90+$angle]" info:`
echo $angle2
convert img2.png mask2.png \
-alpha off -compose copy_opacity -composite \
-background none -rotate "$angle2" +repage -trim +repage \
label_rotated.png

```

If tmp.png has a big black hole inside the label, then the fuzz value may need to be increased and/or the area-threshold needs to be increased to remove a larger black region inside the white label.  Or in place of or addition to the fuzz value, the color of the label in your example might be different enough that the srgb color might need changing. The fuzz value is used to allow the threshold to work on a range of colors near the color specified.

**Reply by lm-sudo** [reply]
Ahh yes, you're right. the color threshold needed to be adjusted. After adjustment: 
![label_rotated3](https://github.com/ImageMagick/ImageMagick/assets/89110072/53de65ef-dc21-4935-9033-12c1268d4d42)

Thank you! You have answered my question so I've marked your response as the answer. This is aside from the original question asked, but say there were more than one color of interest: for example a label that wasn't entirely black but instead had a second color on it like red. Is there anyway to maintain the second color by having a second color threshold or something like that? Or would it require an entirely different approach?

**Comment by fmw42** [comment]
Here is an alternate Imagemagick 6 script.  It is multiple commands and include the use of my script unrotate2, which you will have to get from http://www.fmwconcepts.com/imagemagick

It does a color threshold by a different method using -fuzz XX% -fill white -opaque the_label_color -fill black +opaque white.  Then it removes white or black regions less than 100,000 pixels in area and saves the mask image. Then it gets the cropbox from the mask.  Then it crops the mask and crops the original image. Then it uses my script unrotate2 to get the derotation angle. Then it puts the cropped mask into the alpha channel of the cropped input and derogates them using the denotation angle an finally trims any remaining excess transparent background. 

Input:

![label](https://github.com/ImageMagick/ImageMagick/assets/13458218/91284afa-2407-466c-ab42-0475d221402e)


```
convert label.jpg -auto-orient +repage +write img.png \
-fuzz 4.5% -fill white -opaque "srgb(153,192,233)" -fill black +opaque white \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \
mask.png
cropbox=`convert mask.png -format "%@" info:`
convert mask.png -crop $cropbox +repage mask2.png
convert img.png -crop $cropbox +repage img2.png
angle=`unrotate2 -l 10 mask2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`
declare `convert img2.png -format "aspect_flag=%[fx:w/h<1?1:0]" info:`
[ $aspect_flag -eq 1 ] && angle2=`convert xc: -format "%[fx:90+$angle]" info:`
convert img2.png mask2.png \
-alpha off -compose copy_opacity -composite \
-background none -rotate "$angle2" +repage -trim +repage \
label_rotated.png

```

If you do not want to output all the intermediate images, then add the following line after the output image line.

`rm -f mask.png mask2.png img.png img2.png`

Result:

 
![label_rotated](https://github.com/ImageMagick/ImageMagick/assets/13458218/4a37720b-9f20-4f2f-bc80-db9b96dee3a9)




There is no guarantee about the orientation. It could be off by 90 or 180 or 270 degrees.  Also I am not sure how accurate and reliable it is. It likely will depend upon how good the label edges are in the binary mask image.

**Reply by lm-sudo** [reply]
I'll try using the image magick 6 version with your rotate2 function and see how it works out.

**Comment by fmw42** [comment]
I modified the script to process your second image. I added some morphology to help clean up the thresholded image. I also increased the unrotate2 -l argument from 10 to 20.  You may need to do this if you do not get a reasonable unrotate result. It might need to go up or down. I also fixed the aspect_flag for an error I made that did not affect the first image, but did not work for the second image.

Input:

![label2](https://github.com/ImageMagick/ImageMagick/assets/13458218/5d3cbb9a-efeb-4abd-8020-6bc361c64ef4)


```
convert label2.jpg -auto-orient +repage +write img.png \
-fuzz 4.5% -fill white -opaque "srgb(153,192,233)" -fill black +opaque white +write thresh.png \
-morphology close disk:1 +write morph.png \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \
mask.png
cropbox=`convert mask.png -format "%@" info:`
convert mask.png -crop $cropbox +repage mask2.png
convert img.png -crop $cropbox +repage img2.png
angle=`unrotate2 -l 20 mask2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`
echo $angle
declare `convert img2.png -format "aspect_flag=%[fx:w/h<1?1:0]" info:`
echo $aspect_flag
[ $aspect_flag -eq 1 ] && angle2=`convert xc: -format "%[fx:90+$angle]" info:` || angle2=$angle
echo $angle2
convert img2.png mask2.png \
-alpha off -compose copy_opacity -composite \
-background none -rotate "$angle2" +repage -trim +repage \
label2_rotated.png

```
Optional:
`rm -f thresh.png morph.png mask.png mask2.png img.png img2.png`


Result:

![label2_rotated](https://github.com/ImageMagick/ImageMagick/assets/13458218/accce5f6-75ef-4d1f-8238-bb0051285152)

**Reply by lm-sudo** [reply]
Haven't been able to get this version with the rotate2 function up yet as I'm having a hard time running the script on my ubuntu. Going to keep trying to get this one to work.

So far I've been using the image magick 7 version of the script you provided earlier.

**Reply by fmw42** [reply]
I think you should change convert to magick in my new script. I think it will be better.

```
magick label2.jpg -auto-orient +repage +write img.png \
-fuzz 4.5% -fill white -opaque "srgb(153,192,233)" -fill black +opaque white +write thresh.png \
-morphology close disk:1 +write morph.png \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \
mask.png
cropbox=`magick mask.png -format "%@" info:`
magick mask.png -crop $cropbox +repage mask2.png
magick img.png -crop $cropbox +repage img2.png
angle=`unrotate2 -l 20 mask2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`
echo $angle
declare `magick img2.png -format "aspect_flag=%[fx:w/h<1?1:0]" info:`
echo $aspect_flag
[ $aspect_flag -eq 1 ] && angle2=`magick xc: -format "%[fx:90+$angle]" info:` || angle2=$angle
echo $angle2
magick img2.png mask2.png \
-alpha off -compose copy_opacity -composite \
-background none -rotate "$angle2" +repage -trim +repage \
label2_rotated.png

```
Optional:
`rm -f thresh.png morph.png mask.png mask2.png img.png img2.png`

**Reply by lm-sudo** [reply]
I think somehow I have botched the installation of unrotate2.
After running the above script:
```

1
magick: Expected operand after operator at '' @ error/fx.c/TranslateExpression/2641.

magick: invalid argument for option '-rotate' '' at CLI arg 10 @ error/operation.c/CLISimpleOperatorImage/3297.

```
So the 1 being echoed is definitely $aspect_flag as I checked via manually doing "echo $aspect_flag". This leads me to believe $angle is empty, especially since it's saying '' is being passed to -rotate as an argument. '' would be $angle2 which is determined by $angle which is determined by unrotate2.  Since no error message is brought up in regards to unrotate2, do you think that means unrotate2 is executing but somehow returning an empty string, or is it silently erroring and I messed up installation?

**Reply by lm-sudo** [reply]
Yeah, I'm getting unrotate2 command not found when trying to do unrotate2 -h, okay time to figure this out.

**Reply by fmw42** [reply]
Did you follow the install suggestions on my home page?  Were you running the unrotate2 stand alone and on what image or in the script above?

**Reply by lm-sudo** [reply]
Yes I followed the suggestions on your home page but I think I messed something up along the way, I am extremely new to linux/ubuntu. I am working on a clean linux build running off WSL, which should have allowed me to use everything out of the box like Bruce Rusk as noted on your "Pointers for use:". I wasn't able to do this however so I followed the pointers for use from start to finish(i thought accurately but maybe not) and still wasn't able to get unrotate2 to work. I was using unrotate2 in the script you provided, that is:

```
magick label2.jpg -auto-orient +repage +write img.png \
-fuzz 4.5% -fill white -opaque "srgb(153,192,233)" -fill black +opaque white +write thresh.png \
-morphology close disk:1 +write morph.png \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \
mask.png
cropbox=`magick mask.png -format "%@" info:`
magick mask.png -crop $cropbox +repage mask2.png
magick img.png -crop $cropbox +repage img2.png
angle=`unrotate2 -l 20 mask2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`
echo $angle
declare `magick img2.png -format "aspect_flag=%[fx:w/h<1?1:0]" info:`
echo $aspect_flag
[ $aspect_flag -eq 1 ] && angle2=`magick xc: -format "%[fx:90+$angle]" info:` || angle2=$angle
echo $angle2
magick img2.png mask2.png \
-alpha off -compose copy_opacity -composite \
-background none -rotate "$angle2" +repage -trim +repage \
label2_rotated.png
```
 and it returned the above errors which I believe suggest unrotate2 is not executing.

 I'm going to ask a buddy who's a sysadmin to help me out with installation and setup of the script.

**Reply by fmw42** [reply]
Test this image with standalone unrotate2

![barn_rot10](https://github.com/ImageMagick/ImageMagick/assets/13458218/df9e1a54-0543-4716-8cf9-108be4bd47a0)

`unrotate2 -f 15 -t none barn_rot10.jpg x.jpg`

![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/8260bbf1-49eb-4175-8a11-055706e1ce34)

What result or messages do you get?

**Reply by lm-sudo** [reply]
I currently have the script entitled: unrotate2 in the same directory as my the images I'm working with. I have edited the script so that 
```
imdir="/usr/local/bin"
PATH="${imdir}:${PATH}"
```
is present beneath the default variables and before any image magick is used. I determined the value for the imdir variable by doing print -a convert and using that path(so that is /usr/local/bin/convert minus /convert).

This is the error returned:
unrotate2: command not found. 

I will try some more tinkering later today, if not then tomorrow morning. I think it's likely I'm missing something obvious.

**Reply by fmw42** [reply]
But you said your were using IM 7.  So is /usr/local/bin where Imagemagick 7 is located?

where magick

or

type magick

**Reply by fmw42** [reply]
cd to your directory with your images and the script. Then try

bash unrotate2 -f 15 -t none barn_rot10.jpg x.jpg

is your script labeled just unrotate2 or unrotate2.sh or some other suffix. If not .sh, try changing the script name to unrotate2.sh

Also be sure to make the script executable.

chmod u+x unrotate2.sh

The script does not need installing.  It just needs to be downloaded and placed where you want. Then follow my pointers.

**Reply by lm-sudo** [reply]
I renamed the file to unrotate2.sh and allowed script to be executable(I had already done this before). However, I did not use the word `bash` before typing `unrotate2 -f 15 -t none barn_rot10.jpg x.jpg`.
After doing so, an image x.jpg was created like so:

![x](https://github.com/ImageMagick/ImageMagick/assets/89110072/c90401f8-14b8-46f6-be3e-3f9db2fc826e)

I did get these messages in the terminal, I don't know if they're relevant at all:
```
unrotate2.sh: line 199: type: unrotate2.sh: not found
dirname: missing operand
Try 'dirname --help' for more information.
basename: missing operand
Try 'basename --help' for more information.
derotation=-10.0028;
```

Now that I know unrotate2 is working, I tried the script from before:
```
magick label2.jpg -auto-orient +repage +write img.png \
-fuzz 4.5% -fill white -opaque "srgb(153,192,233)" -fill black +opaque white +write thresh.png \
-morphology close disk:1 +write morph.png \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \
mask.png
cropbox=`magick mask.png -format "%@" info:`
magick mask.png -crop $cropbox +repage mask2.png
magick img.png -crop $cropbox +repage img2.png
angle=`unrotate2 -l 20 mask2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`
echo $angle
declare `magick img2.png -format "aspect_flag=%[fx:w/h<1?1:0]" info:`
echo $aspect_flag
[ $aspect_flag -eq 1 ] && angle2=`magick xc: -format "%[fx:90+$angle]" info:` || angle2=$angle
echo $angle2
magick img2.png mask2.png \
-alpha off -compose copy_opacity -composite \
-background none -rotate "$angle2" +repage -trim +repage \
label2_rotated.png
```

And received this in terminal: 
```

0

magick: invalid argument for option '-rotate' '' at CLI arg 10 @ error/operation.c/CLISimpleOperatorImage/3297.
```
the angle being returned is still an empty string.

Could it be that angle needs to be set to "`bash unrotate2..." instead of "`unrotate2..."? I will mess around with it some more.

**Comment by fmw42** [comment]
I have modified the code again. This time I added converting the mask to its convex hull, which requires IM 7.  This should make the unrotated more reliable and also make the -l argument more robust so that it might not need changing for most images. 

```
magick label2.jpg -auto-orient +repage +write img.png \
-fuzz 4.5% -fill white -opaque "srgb(153,192,233)" -fill black +opaque white +write thresh.png \
-morphology close disk:1 +write morph.png \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100000 \
-connected-components 8 \
mask.png
hull=`magick mask.png -background black -format "%[convex-hull]" info:`
magick mask.png -fill black -colorize 100 -fill white -draw "polygon $hull" -alpha off hull.png
cropbox=`magick hull.png -format "%@" info:`
magick hull.png -crop $cropbox +repage hull2.png
magick mask.png -crop $cropbox +repage mask2.png
magick img.png -crop $cropbox +repage img2.png
angle=`unrotate2 -l 30 hull2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`
echo $angle
declare `magick img2.png -format "aspect_flag=%[fx:w/h<1?1:0]" info:`
echo $aspect_flag
[ $aspect_flag -eq 1 ] && angle2=`magick xc: -format "%[fx:90+$angle]" info:` || angle2=$angle
echo $angle2
magick img2.png mask2.png \
-alpha off -compose copy_opacity -composite \
-background none -rotate "$angle2" +repage -trim +repage \
label_rotated2.png

```
Optional:

`rm -f thresh.png morph.png mask.png hull.png hull2.png mask2.png img.png img2.png`


It worked for me on both your test images.

**Comment by snibgo** [comment]
> This is the error returned:
unrotate2: command not found.

Try with a `./` prefix to `unrotate`, like this:
```
bash ./unrotate2 -f 15 -t none barn_rot10.jpg x.jpg
```

**Reply by fmw42** [reply]
Yes, that is a possibility, too.

**Comment by fmw42** [comment]
These messages 


unrotate2.sh: line 199: type: unrotate2.sh: not found
dirname: missing operand
Try 'dirname --help' for more information.
basename: missing operand
Try 'basename --help' for more information.

Are because you did not put unrotate2 in your path. They won’t hurt. But that means the script cannot find itself for getting the name and directory.

So you do need bash in front of unrotate2.sh. Also you if you are using unrotate2.sh, then the call to it must also use .sh

angle=`bash unrotate2.sh -l 20 mask2.png null: 2>&1 | grep "derotation" | sed -n 's/^derotation=\([0-9.-]*\);$/\1/p'`

Fred

**Reply by lm-sudo** [reply]
OH, I didn't understand that the unrotate2 script had to reside WITHIN /usr/local/bin. As you can tell I'm very green. Now unrotate2 is behaving as expected and successfully rotates and crops the image. Thank you for your help.

By the way, do you expect the image magick 6 script to function more consistently/better than the imagemagick 7 script? I notice it's longer and I assume that's because it is functionally doing more.

**Reply by fmw42** [reply]
They should behave similarly.  IM 7 is a bit slower than IM 6 because IM 7 is HDRI by default when default IM 6 is not. 

There are differences in IM 7 and IM 6 and all my scripts have not been upgraded and tested in IM 7. So if a script does not run properly, then notify me and I will investigate and fix it.

**Reply by fmw42** [reply]
You do not have to have the script in your $PATH, however it is recommended.  Otherwise, you need to preface the script with bash and put the location of Imagemagick in the script.
